File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/conversation/serialization/Movie.kt
package com.xebia.functional.xef.conversation.serialization

import com.xebia.functional.xef.conversation.Conversation
import com.xebia.functional.xef.conversation.llm.openai.OpenAI
import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.store.LocalVectorStore
import kotlinx.serialization.Serializable

@Serializable data class Movie(val title: String, val genre: String, val director: String)

suspend fun main() {
  val openAI = OpenAI()
  //    val openAI = OpenAI(host = "http://localhost:8081/")
  val model = openAI.DEFAULT_SERIALIZATION

  val scope = Conversation(LocalVectorStore(openAI.DEFAULT_EMBEDDING))

  model
    .prompt(
      Prompt("Please provide a movie title, genre and director for the Inception movie"),
      scope,
      Movie.serializer()
    )
    .let { movie ->
      println("The movie ${movie.title} is a ${movie.genre} film directed by ${movie.director}.")
    }
}

File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/conversation/serialization/Movie.scala
package com.xebia.functional.xef.scala.conversation.serialization

import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.scala.conversation.*
import io.circe.Decoder

private final case class Movie(title: String, genre: String, director: String) derives SerialDescriptor, Decoder

@main def runMovie: Unit =
  conversation {
    val movie = prompt[Movie](Prompt("Inception movie genre and director."))
    println(s"The movie ${movie.title} is a ${movie.genre} film directed by ${movie.director}.")
  }
